tests/test_metadata.py

```diff
@@ -0,0 +1,162 @@
+            "signatures": [cls.signature_dict],
+        }
+        cls.pae = b"DSSEv1 29 http://example.com/HelloWorld 11 hello world"
+
+    def test_envelope_from_to_dict(self):
+        """Test envelope to_dict and from_dict methods."""
+
+        envelope_dict = copy.deepcopy(self.envelope_dict)
+
+        # create envelope object from its dict.
+        envelope_obj = Envelope.from_dict(envelope_dict)
+        for signature in envelope_obj.signatures:
+            self.assertIsInstance(signature, Signature)
+
+        # Assert envelope dict created by to_dict will be equal.
+        self.assertDictEqual(self.envelope_dict, envelope_obj.to_dict())
+
+        # Assert TypeError on invalid signature.
+        envelope_dict["signatures"] = [""]
+        with self.assertRaises(FormatError):
+            Envelope.from_dict(envelope_dict)
+
+    def test_envelope_eq_(self):
+        """Test envelope equality."""
+
+        envelope_obj = Envelope.from_dict(copy.deepcopy(self.envelope_dict))
+
+        # Assert that object and None will not be equal.
+        self.assertNotEqual(None, envelope_obj)
+
+        # Assert a copy of envelope_obj will be equal to envelope_obj.
+        envelope_obj_2 = copy.deepcopy(envelope_obj)
+        self.assertEqual(envelope_obj, envelope_obj_2)
+
+        # Assert that changing the "payload" will make the objects not equal.
+        envelope_obj_2.payload = b"wrong_payload"
+        self.assertNotEqual(envelope_obj, envelope_obj_2)
+        envelope_obj_2.payload = envelope_obj.payload
+
+        # Assert that changing the "payload_type" will make the objects not equal.
+        envelope_obj_2.payload_type = "wrong_payload_type"
+        self.assertNotEqual(envelope_obj, envelope_obj_2)
+        envelope_obj_2.payload = envelope_obj.payload
+
+        # Assert that changing the "signatures" will make the objects not equal.
+        sig_obg = Signature("", self.signature_dict["sig"])
+        envelope_obj_2.signatures = [sig_obg]
+        self.assertNotEqual(envelope_obj, envelope_obj_2)
+
+    def test_preauthencoding(self):
+        """Test envelope Pre-Auth-Encoding."""
+
+        envelope_obj = Envelope.from_dict(copy.deepcopy(self.envelope_dict))
+
+        # Checking for Pre-Auth-Encoding generated is correct.
+        self.assertEqual(self.pae, envelope_obj.pae())
+
+    def test_sign_and_verify(self):
+        """Test for creating and verifying DSSE signatures."""
+
+        # Create an Envelope with no signatures.
+        envelope_dict = copy.deepcopy(self.envelope_dict)
+        envelope_dict["signatures"] = []
+        envelope_obj = Envelope.from_dict(envelope_dict)
+
+        key_list = []
+        for key_dict in self.key_dicts:
+            # Test for invalid scheme.
+            valid_scheme = key_dict["scheme"]
+            key_dict["scheme"] = "invalid_scheme"
+            signer = SSlibSigner(key_dict)
+            with self.assertRaises((FormatError, UnsupportedAlgorithmError)):
```

Comment:
does it really raise UnsupportedAlgorithmError?

Reply:
`self.key_dicts` contains RSA, ECDSA, and ED25519 keys. Two of the key raises `FormatError`, and one of them raises `UnsupportedAlgorithmError` for invalid scheme.

Reply:
right :joy: